dbms/src/Databases/DatabaseMySQL.cpp

```diff
@@ -0,0 +1,398 @@
+        { std::make_shared<DataTypeUInt8>(),    "is_nullable" },
+        { std::make_shared<DataTypeUInt8>(),    "is_unsigned" },
+        { std::make_shared<DataTypeUInt64>(),   "length" },
+    };
+
+    WriteBufferFromOwnString query;
+    query << "SELECT "
+             " TABLE_NAME AS table_name,"
+             " COLUMN_NAME AS column_name,"
+             " DATA_TYPE AS column_type,"
+             " IS_NULLABLE = 'YES' AS is_nullable,"
+             " COLUMN_TYPE LIKE '%unsigned' AS is_unsigned,"
+             " CHARACTER_MAXIMUM_LENGTH AS length"
+             " FROM INFORMATION_SCHEMA.COLUMNS"
+             " WHERE TABLE_SCHEMA = " << quote << mysql_database_name
+          << " AND TABLE_NAME IN " << toQueryStringWithQuote(tables_name) << " ORDER BY ORDINAL_POSITION";
+
+    const auto & external_table_functions_use_nulls = global_context.getSettings().external_table_functions_use_nulls;
+    MySQLBlockInputStream result(mysql_pool.Get(), query.str(), tables_columns_sample_block, DEFAULT_BLOCK_SIZE);
+    while (Block block = result.read())
+    {
+        size_t rows = block.rows();
+        for (size_t i = 0; i < rows; ++i)
+        {
+            String table_name = (*block.getByPosition(0).column)[i].safeGet<String>();
+            tables_and_columns[table_name].emplace_back((*block.getByPosition(1).column)[i].safeGet<String>(),
+                getDataType(
+                    (*block.getByPosition(2).column)[i].safeGet<String>(),
+                    (*block.getByPosition(3).column)[i].safeGet<UInt64>() && external_table_functions_use_nulls,
+                    (*block.getByPosition(4).column)[i].safeGet<UInt64>(),
+                    (*block.getByPosition(5).column)[i].safeGet<UInt64>()));
+        }
+    }
+    return tables_and_columns;
+}
+
+void DatabaseMySQL::shutdown()
+{
+    std::map<String, MySQLStorageInfo> tables_snapshot;
+    {
+        std::lock_guard lock(mutex);
+        tables_snapshot = local_tables_cache;
+    }
+
+    for (const auto & table_snapshot : tables_snapshot)
+        table_snapshot.second.storage->shutdown();
+
+    std::lock_guard lock(mutex);
+    local_tables_cache.clear();
+}
+
+void DatabaseMySQL::cleanOutdatedTables()
+{
+    setThreadName("MySQLDBCleaner");
+
+    std::unique_lock lock{mutex};
+
+    const auto quit_requested = [this] { return quit.load(std::memory_order_relaxed); };
+
+    while (!quit_requested())
+    {
+        for (auto iterator = outdated_tables.begin(); iterator != outdated_tables.end();)
+        {
+            if (!iterator->unique())
+                ++iterator;
+            else
+            {
+                const auto table_lock = (*iterator)->lockAlterIntention(RWLockImpl::NO_QUERY);
+
+                (*iterator)->shutdown();
+                (*iterator)->is_dropped = true;
+                iterator = outdated_tables.erase(iterator);
+            }
+        }
+
+        cond.wait_for(lock, cleaner_sleep_time, quit_requested);
```

Comment:
You can move it in a while condition, making `quit_requested` lambda unneeded.

Reply:
This might make the code less readable ?

Reply:
But now it looks redundant.